src/main/java/de/tum/in/www1/artemis/web/rest/LtiResource.java

```diff
@@ -148,13 +152,22 @@ public void launch(@ModelAttribute LtiLaunchRequestDTO launchRequest, @PathVaria
 
         // Note: The following redirect URL has to match the URL in user-route-access-service.ts in the method canActivate(...)
 
-        String redirectUrl = request.getScheme() + // "https"
-                "://" +                                // "://"
-                request.getServerName() +              // "artemis.ase.in.tum.de"
-                (request.getServerPort() != 80 && request.getServerPort() != 443 ? ":" + request.getServerPort() : "") + "/courses/"
-                + exercise.getCourseViaExerciseGroupOrCourseMember().getId() + "/exercises/" + exercise.getId() + (isNewUser ? "?welcome" : "")
-                + (!SecurityUtils.isAuthenticated() ? "?login" : "") + (isNewUser || !SecurityUtils.isAuthenticated() ? "&" : "") + "jwt=" + jwt;
+        UriComponentsBuilder redirectUrlComponentsBuilder = UriComponentsBuilder.newInstance().scheme(request.getScheme()).host(request.getServerName());
+        if (request.getServerPort() != 80 && request.getServerPort() != 443) {
+            redirectUrlComponentsBuilder.port(request.getServerPort());
+        }
+        redirectUrlComponentsBuilder.pathSegment("courses").pathSegment(exercise.getCourseViaExerciseGroupOrCourseMember().getId().toString()).pathSegment("exercises")
+                .pathSegment(exercise.getId().toString());
+
+        if (isNewUser) {
+            redirectUrlComponentsBuilder.queryParam("welcome", "");
```

Comment:
I feel like empty queryParams are a bit weird, but I guess this is common practice for some reason? Never seen it before

Reply:
I tried null, but the method complained. If you know another way to avoid the `=` sign, I'm happy to know it